tests/cpp/test_multidevice_transformer.cpp

```diff
@@ -600,6 +587,10 @@ TEST_P(DistributedTransformerTest, Forward) {
     tv->setDeviceMesh(mesh);
   }
 
+  shardBetween({mha_in->definition()}, {mha_out->definition()}, mha_w0);
+  shardBetween({mlp_in->definition()}, {mlp_out->definition()}, mlp_w0);
```

Comment:
```suggestion
  shardBetween({mha_in}, {mha_out}, mha_w0);
  shardBetween({mlp_in}, {mlp_out}, mlp_w0);
```
is this any different? If it's indeed equivalent, we could simply remove the Expr-version of shardBetween.

Reply:
There is an equivalent version, it's just sometimes easier to think of it through expression boundaries. I can remove if that helps merge the PR.

Reply:
> it's just sometimes easier to think of it through expression boundaries. 

Why? Because the TensorView version has a weird from-is-inclusive-but-to-is-exclusive semantics? 

> I can remove if that helps merge the PR.

Up to you -- you've been on the front line of this and it helps to make yourself comfortable!

Reply:
It makes it slightly easier to get boundaries correct. For example, if you have any expression with multiple inputs like c = matmul(a, b) and b is part of the `to` tensorviews but not a, then the propagation will visit a's and its producers, which may or may not be expected. 

Using the expression version helps to mimic the between behavior, though it doesn't help with non-linear regions like residuals.